### DISCUSSION TITLE
Multi-process support for Android

### DISCUSSION THREAD

**Discussion by TedaLIEz** [microsoft/cpp_client_telemetry#1158] [General]
# Is your question about specific 1DS SDK APIs or Features?

Our team is working on a multi-process design of our Android application, I'm wondering if cpp_client_telemetry SDK is suitable for multi-process scenario. 

For example, can we acquire multiple logger instances in each process, and those logger instances shall emit telemetry to backend in different tenants? 

In multi-process scenario, it is hard to deal with synchronization because mutex is no longer working between processes. Besides, we don't want to pay the price to use IPC to emit telemetry. So, the best option for us is that each process should emit telemetry on its own.

**Comment by lalitb** [comment]
> For example, can we acquire multiple logger instances in each process, and those logger instances shall emit telemetry to backend in different tenants?

I am not Android expert, but if these are the OS processes, each of these process need to create a separate LogManager instance, which further creates a tenant specific logger.

**Comment by anod** [comment]
We use multiple instances (2-3 android, 3-4 iOS) with different configurations and tenants, every tenant store data in a separate db

As Lalit mention you should use `LogManagerProvider.createLogManager(config)`, also provide unique `CFG_STR_FACTORY_NAME`

Example:
```kotlin
val config = LogManager.logConfigurationFactory().apply {
    set(LogConfigurationKey.CFG_STR_PRIMARY_TOKEN, token)
    set(LogConfigurationKey.CFG_STR_COLLECTOR_URL, collectorUrl)
    set(LogConfigurationKey.CFG_STR_FACTORY_NAME, "PBI_OneAuth_Logger" + UUID.randomUUID())
}

val logManager = LogManagerProvider.createLogManager(config)
 ```

**Reply by TedaLIEz** [reply]
@anod Do we need to create multiple LogManager instances for different processes before calling `LogManager.initialize(String tenantToken, ILogConfiguration configuration)` multiple times?

**Reply by anod** [reply]
No need to call initialize, we keep reference to ILogger

1. Call to `this.logger = logManager.getLogger(token, "", "")`
2. Then log multiple events using logger `this.logger.logEvent(EventProperties(data.name, map))`

**Reply by TedaLIEz** [reply]
@anod Oh, I didn't notice this API, so do you mean (for example):

```
// process A
val config = LogManager.logConfigurationFactory().apply {
    set(LogConfigurationKey.CFG_STR_PRIMARY_TOKEN, token)
    set(LogConfigurationKey.CFG_STR_COLLECTOR_URL, collectorUrl)
    set(LogConfigurationKey.CFG_STR_FACTORY_NAME, "PBI_OneAuth_Logger" + UUID.randomUUID())
}

val logManager = LogManagerProvider.createLogManager(config)
this.logger = logManager.getLogger(tokenA, "")
```


```
// process B
val config = LogManager.logConfigurationFactory().apply {
    set(LogConfigurationKey.CFG_STR_PRIMARY_TOKEN, token)
    set(LogConfigurationKey.CFG_STR_COLLECTOR_URL, collectorUrl)
    set(LogConfigurationKey.CFG_STR_FACTORY_NAME, "PBI_OneAuth_Logger" + UUID.randomUUID())
}

val logManager = LogManagerProvider.createLogManager(config)
this.logger = logManager.getLogger(tokenB, "")
```

Is this correct? BTW I previously used `LogManager.initialize(String tenantToken, ILogConfiguration configuration)` to get a logger instance, what's the difference between this API and `LogManager#getLogger`?

**Reply by TedaLIEz** [reply]
Hi @anod @lalitb Could you please help look at my demo code here? What's the difference between `logManager.getLogger` and ` LogManager.initialize(String tenantToken, ILogConfiguration configuration)`, both APIs will return a logger instance.

**Reply by anod** [reply]
The code looks ok, I am away right now and don't remember what exactly the difference,  afaik initialize not suitable for multiple loggers, it returns existing logger

**Reply by TedaLIEz** [reply]
Yes, looks like `initialize` always returns the same logManager instance, but it returns different logger instances if tenant token is different. However I'm not clear about having multiple logManager instances vs one unique logManager

**Reply by TedaLIEz** [reply]
Ok, I'll use the following code for initializing:

```
val httpClient = HttpClient(context.applicationContext)
OfflineRoom.connectContext(context.applicationContext)
val logConfiguration = LogManager.logConfigurationFactory()
logConfiguration[LogConfigurationKey.CFG_STR_COLLECTOR_URL] = endPointUrl
logConfiguration[LogConfigurationKey.CFG_STR_FACTORY_NAME] = endPointUrl
logConfiguration[LogConfigurationKey.CFG_STR_PRIMARY_TOKEN] = tenantToken
val logManager = LogManagerProvider.createLogManager(logConfiguration)
mLogger = logManager.getLogger(tenantToken, "", "")  // declare logger as a member in class
```

Does this work for multiple logger instances with different tenant in multi-processing?
